### PR TITLE
Ddcnl 4653

### PR DESCRIPTION
**Pull request by brendanstraughan**

### CODE REVIEW

File: app/controllers/AtsMergePageController.scala [hunk 1]
```diff
@@ -58,15 +58,20 @@ class AtsMergePageController @Inject()(
 
     atsMergePageService.getSaAndPayeYearList.map {
       case Right(atsMergePageViewModel) =>
-        Ok(
-          atsMergePageView(
-            atsMergePageViewModel,
-            formWithErrors.getOrElse(atsForms.atsYearFormMapping),
+        val actingAsAttorneyFor: Option[ActingAsAttorneyFor] = atsMergePageViewModel.saData.utr.isEmpty match {
+          case true => None
+          case false =>
```
Comment (pascal22p) [inline app/controllers/AtsMergePageController.scala:63]: This should be a if statement.

File: app/controllers/AtsMergePageController.scala [hunk 2]
```diff
@@ -58,15 +58,20 @@ class AtsMergePageController @Inject()(
...
             getActingAsAttorneyFor(
               request,
               atsMergePageViewModel.saData.forename,
               atsMergePageViewModel.saData.surname,
               atsMergePageViewModel.saData.utr)
```
Comment (pascal22p) [inline app/controllers/AtsMergePageController.scala:68]: Add your check for utr in getActingAsAttorneyFor. You already return an option so it would be more logical to have the check in there.

File: app/services/AtsListService.scala [hunk 3]
```diff
@@ -163,9 +165,13 @@ class AtsListService @Inject()(
       case (Left(_), identifier) =>
         auditService.sendEvent(
           AuditTypes.Tx_FAILED,
-          Map(
-            "userId"         -> request.userId,
-            "userIdentifier" -> identifier.value
-          ))
+          Map("userId" -> request.userId, "userIdentifier" -> {
+            identifier match {
+              case Some(x: TaxIdentifier) => x.value
```
Comment (struong) [inline app/services/AtsListService.scala:170]: rename x please

File: app/services/AtsListService.scala [hunk 4]
```diff
@@ -163,9 +165,13 @@ class AtsListService @Inject()(
       case (Left(_), identifier) =>
         auditService.sendEvent(
           AuditTypes.Tx_FAILED,
```
Comment (struong) [inline app/services/AtsListService.scala:167]: Its not a failure if this user does not have a UTR in this scenario, you should leave it as a `Tx_SUCCEEDED`

File: app/services/AtsService.scala [hunk 5]
```diff
@@ -90,9 +90,10 @@ class AtsService @Inject()(
     val requestedUTR = authUtils.getRequestedUtr(account, agentToken)
 
     //This warning is unchecked because we know that AuthorisedFor will only give us those accounts
-    val gotData = (account: @unchecked) match {
-      case agentUar: Uar        => middleConnector.connectToAtsOnBehalfOf(agentUar, requestedUTR, taxYear)
-      case individualUtr: SaUtr => middleConnector.connectToAts(individualUtr, taxYear)
+    val gotData = (account: @unchecked, requestedUTR) match {
```
Comment (struong) [inline app/services/AtsService.scala:93]: duplicated code

File: app/services/AtsService.scala [hunk 6]
```diff
@@ -134,7 +135,7 @@ class AtsService @Inject()(
             "time"      -> new Date().toString
           )
         )
-      case _: SaUtr =>
+      case Some(_: SaUtr) =>
```
Comment (struong) [inline app/services/AtsService.scala:138]: What happens on None?

File: app/utils/AccountUtils.scala [hunk 7]
```diff
@@ -17,18 +17,23 @@
...
 import play.api.i18n.Messages
 import play.api.mvc.Request
 import uk.gov.hmrc.domain.{SaUtr, TaxIdentifier, Uar}
 
 trait AccountUtils {
-  def getAccount(request: AuthenticatedRequest[_]): TaxIdentifier =
-    request.agentRef.getOrElse(request.saUtr.getOrElse(SaUtr("")))
+  def getAccount(request: AuthenticatedRequest[_]): Option[TaxIdentifier] =
+    request match {
+      case x if (x.agentRef.isDefined) => Some(x.agentRef.get)
```
Comment (struong) [inline app/utils/AccountUtils.scala:28]: x is request so you are not really matching on anything here, can you see a nicer way of writing this?

File: app/utils/AccountUtils.scala [hunk 8]
```diff
@@ -17,18 +17,23 @@
...
+      case x if (x.saUtr.isDefined)    => Some(x.saUtr.get)
+      case _                           => Some(SaUtr(""))
+    }
+
   //This warning is unchecked because we know that AuthorisedFor will only give us those accounts
   def getAccountId(request: AuthenticatedRequest[_]): String = (getAccount(request): @unchecked) match {
-    case sa: SaUtr => sa.utr
-    case ta: Uar   => ta.uar
+    case Some(sa: SaUtr) => sa.utr
+    case Some(ta: Uar)   => ta.uar
```
Comment (struong) [inline app/utils/AccountUtils.scala:36]: What happens when its None?

File: app/utils/AuthorityUtils.scala [hunk 9]
```diff
@@ -25,31 +25,31 @@ class AuthorityUtils @Inject()() {
 
   def checkUtr(utr: String, agentToken: Option[AgentToken])(implicit request: AuthenticatedRequest[_]): Boolean =
     (AccountUtils.getAccount(request), agentToken) match {
-      case (agentAccount, None) if (AccountUtils.isAgent(request)) =>
+      case (Some(agentAccount), None) if (AccountUtils.isAgent(request)) =>
```
Comment (struong) [inline app/utils/AuthorityUtils.scala:28]: What happens when its None?

File: test/services/AuthorityUtilsSpec.scala [hunk 10]
```diff
@@ -116,24 +116,22 @@ class AuthorityUtilsSpec extends BaseSpec {
 
     "return the utr when user account with no agent token" in new TestService {
       val result = getRequestedUtr(account, None)
```
Comment (struong) [inline test/services/AuthorityUtilsSpec.scala:118]: Missing tests when account is None

**Review by struong — COMMENTED** [review]
You are missing tests for the majority of your changes when the SaUTR and UAR is None

File: app/services/AtsService.scala (hunk 6 above)
Comment (brendanstraughan) [reply app/services/AtsService.scala:138]: This cannot be None. account comes from     val account = utils.AccountUtils.getAccount(request)
 which does not return a None

File: app/utils/AuthorityUtils.scala [hunk 11]
```diff
@@ -25,31 +25,31 @@ class AuthorityUtils @Inject()() {
...
           if (taxsAgent == Uar(agentToken.agentUar)) {
             SaUtr(agentToken.clientUtr)
           } else {
             throw AgentTokenException(s"Incorrect agent UAR: ${taxsAgent.uar}, ${agentToken.agentUar}")
           }
         }
-      case sa: SaUtr => sa
+      case Some(sa: SaUtr) => Some(sa)
+
+      case _ => None
```
[the remark this answers is not in the document]
Comment (brendanstraughan) [reply app/utils/AuthorityUtils.scala:53]: We can remove this as getAccount will not return None

File: app/services/AtsListService.scala (hunk 3 above)
Comment (Sachin300793) [reply app/services/AtsListService.scala]: done

File: app/services/AtsListService.scala [hunk 12]
```diff
@@ -140,18 +141,18 @@ class AtsListService @Inject()(
       data.get
     }
 
-  private def sendAuditEvent(account: TaxIdentifier, dataOpt: Either[Int, AtsListData])(
+  private def sendAuditEvent(account: Option[TaxIdentifier], dataOpt: Either[Int, AtsListData])(
```
Comment (struong) [inline app/services/AtsListService.scala:144]: Missing coverage for when account is `None`

File: app/services/AtsService.scala (hunk 6 above)
Comment (struong) [reply app/services/AtsService.scala:138]: You need to handle the None case or you will get a pattern match exception

File: app/utils/AccountUtils.scala [hunk 13]
```diff
@@ -17,18 +17,26 @@
...
+      None
+    }
+
   //This warning is unchecked because we know that AuthorisedFor will only give us those accounts
   def getAccountId(request: AuthenticatedRequest[_]): String = (getAccount(request): @unchecked) match {
-    case sa: SaUtr => sa.utr
-    case ta: Uar   => ta.uar
+    case Some(sa: SaUtr) => sa.utr
+    case Some(ta: Uar)   => ta.uar
+    case _ => ""
```
Comment (struong) [inline app/utils/AccountUtils.scala:39]: This is confusing, you had this as an Option before and now it's an empty string? Keep it as an Option rather than pass a String type around. It will simplify your code

File: app/utils/AuthorityUtils.scala [hunk 14]
```diff
@@ -25,31 +25,31 @@ class AuthorityUtils @Inject()() {
...
-      case (agentAccount, None) if (AccountUtils.isAgent(request)) =>
+      case (Some(agentAccount), None) if AccountUtils.isAgent(request) =>
         true
-      case (agentAccount, Some(agentToken)) if (AccountUtils.isAgent(request)) =>
+      case (Some(agentAccount), Some(agentToken)) if (AccountUtils.isAgent(request)) =>
         SaUtr(utr) == SaUtr(agentToken.clientUtr)
-      case (account: SaUtr, _) =>
+      case (Some(account: SaUtr), _) =>
         SaUtr(utr) == account
+      case _ => false
```
Comment (struong) [inline app/utils/AuthorityUtils.scala:34]: Missing tests

File: test/services/AtsListServiceSpec.scala [hunk 15]
```diff
@@ -467,7 +508,7 @@ class AtsListServiceSpec extends BaseSpec {
 
         when(mockAuthUtils.checkUtr(any[String], any[Option[AgentToken]])(any[AuthenticatedRequest[_]]))
           .thenReturn(false)
-        when(mockAuthUtils.getRequestedUtr(any[TaxIdentifier], any[Option[AgentToken]]))
+        when(mockAuthUtils.getRequestedUtr(any[Option[TaxIdentifier]], any[Option[AgentToken]]))
```
Comment (struong) [inline test/services/AtsListServiceSpec.scala:511]: This can no longer throw

File: test/controllers/AtsMergePageControllerSpec.scala [hunk 16]
```diff
@@ -115,6 +115,31 @@ class AtsMergePageControllerSpec extends ControllerBaseSpec with ScalaFutures wi
...
+      val agentRequest = AuthenticatedRequest(
+        "userId",
+        None,
+        Some(SaUtr(testUtr)),
+        None,
+        true,
+        true,
+        ConfidenceLevel.L50,
+        fakeCredentials,
+        FakeRequest("GET", controllers.routes.AtsMergePageController.onPageLoad + "")
```
Comment (struong) [inline test/controllers/AtsMergePageControllerSpec.scala:129]: Whats this bit for? `+ ""`

File: app/utils/AuthorityUtils.scala [hunk 17]
```diff
@@ -25,31 +25,31 @@ class AuthorityUtils @Inject()() {
...
+        agentToken map { agentToken =>
           if (taxsAgent == Uar(agentToken.agentUar)) {
             SaUtr(agentToken.clientUtr)
           } else {
             throw AgentTokenException(s"Incorrect agent UAR: ${taxsAgent.uar}, ${agentToken.agentUar}")
           }
         }
-      case sa: SaUtr => sa
+      case Some(sa: SaUtr) => Some(sa)
+      case _               => None
```
Comment (struong) [inline app/utils/AuthorityUtils.scala:53]: Don't need this None line here, this is just a map

File: app/utils/AccountUtils.scala [hunk 18]
```diff
@@ -17,18 +17,26 @@
...
-import models.ActingAsAttorneyFor
+import models.{ActingAsAttorneyFor, AgentToken}
 import play.api.i18n.Messages
 import play.api.mvc.Request
 import uk.gov.hmrc.domain.{SaUtr, TaxIdentifier, Uar}
 
 trait AccountUtils {
-  def getAccount(request: AuthenticatedRequest[_]): TaxIdentifier =
-    request.agentRef.getOrElse(request.saUtr.getOrElse(SaUtr("")))
+  def getAccount(request: AuthenticatedRequest[_]): Option[TaxIdentifier] =
```
Comment (struong) [inline app/utils/AccountUtils.scala:26]: missing tests

File: app/services/AtsListService.scala (hunk 12 above)
Comment (brendanstraughan) [reply app/services/AtsListService.scala:144]: added test AtsListServiceSpec:473

File: test/services/AtsListServiceSpec.scala (hunk 15 above)
Comment (brendanstraughan) [reply test/services/AtsListServiceSpec.scala:511]: Changed exception message

File: app/utils/AccountUtils.scala (hunk 18 above)
Comment (brendanstraughan) [reply app/utils/AccountUtils.scala:26]: AuthorityUtilsSpec:115 will cover this .getAccount None Scenario and there a few other tests in there aswell

**Review by pascal22p — APPROVED** [review]
